bootstrap: upgrade to v2.3.2.

amanagr opened the pull request:
I don't think we can override the changes in https://github.com/zulip/zulip/commit/de4a22d94ba7b52f0c773afcbe45e9c2ef514261 in zulip.scss, so we will have to make change in the file itself. Is that fine?

andersk commented:
I’d very much prefer to use 100% stock Bootstrap directly from NPM, delete the copy from our tree, and resist the temptation to fork it again at all. The main [bootstrap package](https://www.npmjs.com/package/bootstrap) doesn’t have version 2.3.2, but there’s a [bootstrap-2.3.2 package](https://www.npmjs.com/package/bootstrap-2.3.2).

amanagr commented:
> I’d very much prefer to use 100% stock Bootstrap directly from NPM, delete the copy from our tree, and resist the temptation to fork it again at all. The main bootstrap package doesn’t have version 2.3.2, but there’s a bootstrap-2.3.2 package.

Can we please do it when we upgrade to v3.x since I will have to repeat the process of fixing .btn , .popover, .typeahead and .tooltip if we do it in v2.3.2?

timabbott commented on static/styles/zulip.scss:
```diff
@@ -546,6 +546,69 @@ strong {
+    }
+}
+
+li.actual-dropdown-menu > a:focus {
```
Can you add this sort of comment?
`/* Copied from bootstrap 2.1.1 CSS for dropdown-menu li > a:focus */`

amanagr replied:
Okay!

timabbott commented on static/styles/zulip.scss:
```diff
@@ -546,6 +546,69 @@ strong {
+// .dropdown-menu from v2.3.2
+// + https://github.com/zulip/zulip/commit/7a3a3be7e547d3e8f0ed00820835104867f2433d
+// basic idea of this fix is to remove decorations from :hover and apply them only
+// on :focus.
```
Can't we just use a `!important` override for just the one or two properties we want to override?  That seems likely to be a lot cleaner.

amanagr replied:
I would prefer to keep the whole of `.dropdown-menu` css here to avoid future conflicts when upgrading to v3.x. Most of the fields here are important, I can trim down 1 or 2 but I don't think it would really make a difference.

timabbott commented:
> Can we please do it when we upgrade to v3.x since I will have to repeat the process of fixing .btn , .popover, .typeahead and .tooltip if we do it in v2.3.2?

Yeah, I think that's reasonable; if this process lets us simplify the complexity of understanding the next upgrades, it's well worth doing in this form.  I'm happy to block merging all code changes to bootstrap not related to upgrading it until we can get it from NPM.

> Any commit not mentioned here is not worth worrying about.

Would you be up for listing the commits that we can ignore or explaining why we can not worry about them?  (E.g. they're formatting changes).

> Changes to bootstrap-responsive will be reverted in a future
    commit on v2.3.2. They should be moved in zulip.scss when
    upgrading to v3.x.

Can you extend this paragraph of the commit message to explain what you mean here?  Is the point that we backported some details from newer bootstrap-responsive from v2.3.2 and so we'll get those changes automatically?

> bootstrap-js: Upgrade and reset to v2.1.1.

Can you explain why it's the right approach to fork the tooltip library rather than doing something more tactical?  Is 21ccf45db9a353c99a86ff41296399532cd68edf the issue?  Might be worth including in that commit message a list of all the commit IDs that are not typeahead changes -- that'll be a useful resource for future work to have collected in one place.

And also I don't understand why we seem to replace it entirely in the v2.3.2 commit.  I think it might be cleaner to have the structure for that piece be:
* Extract the v2.3.2 bootstrap version.
* Add commits on top of it re-applying our changes to the tooltip module, so that it's easy to see and keep track of what we've changed (And add very visible comments with ZULIP in them explaining what we've changed and why near the changes).
* Also, I had thought we'd made changes to the bootstrap popover library.

Finally, how confident are you in having ported all the changes forward correctly?

timabbott commented:
This is great progress and a valuable checkpoint that will let us look at a lot less deep history to understand our changes; posted a bunch of comments on how we might do even better before integrating this.

amanagr commented:
bootstrap.scss changes before this PR:
Changes added then removed, so irrelevant -
d7f9a2151970f4c69d1feb394423c180e85ccd43 api docs: Move json-api-example css to portico.scss.
9b740df2778bf9af3c072da3e7d2312b140f044b api_docs: Specify max width for long code examples. Fixes #10673.
> Would you be up for listing the commits that we can ignore or explaining why we can not worry about them? (E.g. they're formatting changes).
Reducing z-index of overlays doesn't make sense - 1143ed721960042b01d1024dc453bbab0ccddb99 redesign: Change /#settings and /#administration to an overlay.
Moved to zulip.scss - 5d0960ae0445e06cb1e4b3b86d33c36ef34fbf6b Change color of inline code from red to black.
Moved to zulip.scss - 27217fddb34e79e2be1cdde65ee63d3de4e5c9a1 make bootstrap inline-code preserve whitespace
Reverted during reset to v2.1.1 and re-applied after reset to v2.3.2 - 2429ed6eebf2f8d713b8cedb82add771f8aab0fa Remove old Bootstrap2 button code.
Reverted during reset to v2.1.1 (It was not a typo but it adds support for IE9) - 3142d74d391d5eab698fba1784c1b2f72660a331 Typo fix.
Reverted during reset to v2.1.1 - ead73f37f97fa76ca6d35d1f055691dc61374a79 Remove Glyphicons completely.
Irrelevant (file moved) - 3bba0cc927c73c05c893f72e0b509de1bb4d8f48 Move zephyr/static to just static.
Moved to zulip.scss - 04df44f91ffb7bba25cfbe3d10548e5e36e80f19 Allow a backdoor for styling dropdown hovers in bootstrap.
Moved to zulip.scss - 7a3a3be7e547d3e8f0ed00820835104867f2433d [third] Fix sticky hover background color in typeaheads.

Since we're doing a reset to bootstrap's version, we want to show their license? - 
441e4296e0ae4678154564ec7dfc181533cb74cd Add copyright notices to third-party files
5e2c493ea55237114024fbe9a79b2f94c5702203 Remove CSS/JS comments that were making their way into minified code

Irrelevant - 89ed4441fdfd4e80ca3deed9b85cb764af90f39d Update to the latest version of bootstrap.
Irrelevant (file moved) - 8f1cee0b882ebb656ee25bbf27564d7664432a39 Move third-party components into static/third/
Irrelevant - 1490ae1855d815270d91c6a42c825ab697e7cee4 Add Bootstrap

amanagr commented:
I will reply to rest of the comments tomorrow, I am not feeling good today.

timabbott commented:
The new version looks great; closing this in favor of it.  I don't think we need the "add license header" commits; we've moved that data to `docs/THIRDPARTY`.
